### PR TITLE
uDiscovery redesign

### PR DESCRIPTION
**Pull request by stevenhartley**
The following PR is a reboot of uDiscovery to be able to take all the lessons learned from existing implementations in production.

#137

### CODE REVIEW

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * SPDX-FileCopyrightText: 2023 Contributors to the Eclipse Foundation
+ * SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:2]: this does not need to be changed, given that it is an existing file and unchanged major version ...

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 2]
```diff
@@ -25,332 +24,97 @@ option java_multiple_files = true;
 // Enables generation of generic service attributes in C++
 option cc_generic_services = true;
 
-// Platform USE Discovery Service Interface
+// Client facing APIs to the uProtocol Discovery Service.
+// 
+// The uDiscovery service is used to store information about devices and services namely their
+// addresses (in URI format), and properties of topics. This API is used by clients (uEs) to be
+// about to access the information within the database. Clients talk to their local uDiscovery service
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:31]:
```suggestion
// addresses (in URI format), and properties of topics. This API is used by clients (uEs) to
// access the information within the database. Clients talk to their local uDiscovery service
```

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 3]
```diff
@@ -25,332 +24,97 @@ option java_multiple_files = true;
...
+// who, when is unable to find what it is looking for, queries the central database to find the information.
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:32]:
```suggestion
// which, if unable to find what it is looking for, queries the central database to find the information.
```

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 4]
```diff
@@ -25,332 +24,97 @@ option java_multiple_files = true;
...
+// UDiscovery works in heiarchial mode where the local uDiscovery service is the first point of contact
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:33]:
```suggestion
// The UDiscovery service instances form a hierarchy where the local uDiscovery service is the first point of contact
```

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 5]
```diff
@@ -25,332 +24,97 @@ option java_multiple_files = true;
...
-  // for a query.
-  // What is returned is a list of UUris that match the query.
-  // If lookup fails, the response will be a UStatus with
-  // 2. code.NOT_FOUND: No matching UUris were found
-  // 3. code.INVALID_ARGUMENT: The passed UUri is invalid
-  // 4. code.PERMISSION_DENIED: The caller does not have permission to perform the query
-  rpc LookupUri(LookupUriRequest) returns (LookupUriResponse) {
+  // Find Services. 
+  //
+  // Discover t authorities, instances, and versions of services based on the passed URI.
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:48]:
```suggestion
  // Discover authorities, instances, and versions of services based on the passed URI.
```

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 6]
```diff
@@ -25,332 +24,97 @@ option java_multiple_files = true;
...
+  // The only mandatory portion of the URI that must be passed is the ue_id, the rest can be discovered.
+  // When you pass the wildcard for authority_name, the search will be done up to the domain, in order to seach
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:50]: this doesn't seem to be in line with the examples provided in the README.
In general, IMHO we should try not to duplicate the information in both this proto3 file and the README. My preference is to provide the detailed docs in the README and only add sparse comments here ...

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 7]
```diff
@@ -25,332 +24,97 @@ option java_multiple_files = true;
...
-}
-
-// The (empty) response message representing the successful execution of the UpdateNode operation.
-message UpdateNodeResponse {}
+// Request to get information about a topic
+// Example might be to get information about a specific topic, all topics for a service (using wildcard resource_id),
+// or all topics for all services for a given device (using wildcard ue_id and resource_id)
+message GetServiceTopicsRequest {
+  // The URI of the topics that we would like to query. for example, 
+  // to fetch a specific topic, we would set
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:76]: this sentence seems to be unfinished ...

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 8]
```diff
@@ -25,332 +24,97 @@ option java_multiple_files = true;
...
+  // The topic metadata
+  UServiceTopic info = 2;
 
-  // How deep in the node tree should the notifications be sent for. A value of -1 or if
-  // the field is not present in the message, signifies that changes to any child nodes will trigger
-  // a Notification event. A value of 0 returns only the parent node. Any other value specified is
-  // the depth of child nodes to receive notifications for.
-  optional int32 depth = 3;
+  // How long the topic metadata is valid for in seconds.
+  // If the metadata is older than this value, the client MAY re-fetch the metadata.
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:100]:
```suggestion
  // If the metadata is older than this value, the client SHOULD re-fetch the metadata.
```

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 9]
```diff
@@ -25,332 +24,97 @@ option java_multiple_files = true;
...
+  // If the field is missed or the value is 0, the client MAY assume the metadata lives forever.
+  optional uint32 ttl = 3;
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:102]: IMHO we do not need to make it optional then. The default value of an int field is `0` by (proto3) definition.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery_replicator.proto [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+option java_package = "org.eclipse.uprotocol.core.udiscovery.v3";
+option java_outer_classname = "UDiscoveryReplicatorProto";
+option java_multiple_files = true;
+ 
+  
+// UDiscovery Replicator is a set of APIs that are used internal to UDiscovery to populate and replicate
+// information in the uDiscovery database. This API is for internal (to UDiscovery service) use only and 
+// not accessible for client uEs to use.
+ service UDiscoveryReplicator {
+  option (uprotocol.service_name) = "core.udiscovery"; // Service name
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery_replicator.proto:31]: FMPOV this should have a name that is different from the `UDiscovery` service defined in the other proto.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery_replicator.proto [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  option (uprotocol.service_version_major) = 3;
+  option (uprotocol.service_version_minor) = 0;
+  option (uprotocol.service_id) = 1;
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery_replicator.proto:34]: and, consequently, a different service ID.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery_replicator.proto [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  // If the metadata has expired, the UDiscovery service that received this data must flush the ServiceTopic info.
+  // If the field is missed, the ServiceTopicInfo is valid forever.
+  // If the field is set to 0, the ServiceTopicInfo should be removed from the database immediately.
+  optional uint32 ttl = 2;
+}
+
+
+// Empty message returned from SetServiceTopics when the command returned successfully
+message SetServiceTopicsResponse {}
+
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery_replicator.proto:66]: missing EOL

File: up-l3/udiscovery/v3/README.adoc [hunk 13]
```diff
@@ -1,11 +1,12 @@
...
-:toc:
+:toc: preamble
 :sectnums:
+:source-highlighter: highlight.js
 
 The key words "*MUST*", "*MUST NOT*", "*REQUIRED*", "*SHALL*", "*SHALL NOT*", "*SHOULD*", "*SHOULD NOT*", "*RECOMMENDED*", "*MAY*", and "*OPTIONAL*" in this document are to be interpreted as described in https://www.rfc-editor.org/info/bcp14[IETF BCP14 (RFC2119 & RFC8174)]
 
 ----
-SPDX-FileCopyrightText: 2023 Contributors to the Eclipse Foundation
+SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/README.adoc:9]: no need to update the year

File: up-l3/udiscovery/v3/service.adoc [hunk 14]
```diff
@@ -0,0 +1,253 @@
+= uDiscovery Service
+:toc: preamble
+:sectnums:
+:source-highlighter: highlight.js
+:client-proto-ref: xref:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto]
+:replicator-proto-ref: xref:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery_replicator.proto]
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:6]: you should only use `xref` for references to AsciiDoc resources. For everything else, use `link`

File: up-l3/udiscovery/v3/service.adoc [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+=== Terms
+
+- *Fully Qualified Service URI (FQURI):* A URI that contains all the information needed to uniquely identify a service. A fully qualified service URI contains the following information:
+  - *ue_id:* The unique identifier of the service (both the instance and service ID portions).
+  - *authority_name:* The name of the authority that hosts the service.
+  - *ue_version_major:* The version of the service.
+  - *resource_id:* The resource ID (value of 0 is used to indicate the default ID for the service).
+
+
+== Client APIs
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:50]: I am surprised to see the client facing API being described here again. Isn't the Client facing API described in the `client.adoc` file? Isn't that the whole purpose of the split-up of the original proto service into two separate (proto) services (`UDiscovery` and `UDiscoveryReplicator`)?

File: up-l3/udiscovery/v3/service.adoc [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+Recursive lookups happen when the client sets the following information in the request message of the FindService() or FindServiceTopics() API. wishes to To recursively lookup up to the domain (or central) levels, the client calling the API :
+[.specitem,oft-sid="dsn~discovery-recursive-authority-wildcard~1",oft-needs="impl,test"]
+--
+*  `authority_name` *MUST* contain the wildcard `*`.
+--
+
+If `search_central` inside of the `FindServicesRequest` or `GetServiceTopicsRequest` is `true``:
+[.specitem,oft-sid="dsn~discovery-recursive-central~1",oft-needs="impl,test"]
+--
+* *MUST* search recursively up to and including the central instance.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:115]: Does a UDiscover instance know whether the _next hop_ already is the central instance? Does it need to know? FMPOV it would be sufficient to require a service instance to ask _the next in chain_, if one is configured ...

File: up-l3/udiscovery/v3/service.adoc [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+--
+
+UDiscovery services perform a recursive search when the local or domain instance calls `FindService()` or `FindServiceTopics()` API and then propagates the response back to the client.
+
+[.specitem,oft-sid="dsn~discovery-recursive-async~1",oft-needs="impl,test"]
+--
+* The original local search *MUST ONLY* return the response message when the recursive search has completed.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:122]: FMPOV we should also shortly comment on using the TTL value appropriately for this purpose ...

File: up-l3/udiscovery/v3/service.adoc [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+--
+
+[.specitem,oft-sid="dsn~discovery-recursive-internal~1",oft-needs="impl,test"]
+--
+* The uDiscovery service *MUST* cache the results of the recursive search for the duration `ttl` that is returned in the response message.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:127]: I believe that the `ttl` being referred to here is the `ttl` contained _in_ the response message _itself_, not the `ttl` contained in the UMessage attributes of the response message?

File: up-l3/udiscovery/v3/service.adoc [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+----
+
+
+== Implementation Consideration
+
+It is possible the information in domain or central instances might become out of sync with what is stored in the other instances. In order to rectify this situation:
+
+[.specitem,oft-sid="dsn~discovery-data-reconciliation~1",oft-needs="impl,test"]
+--
+* Domain and central instances *MUST* ensure that the information within its database is accurate and up to date and reconcile any differences periodically using the client and replication APIs described in this specification.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:214]:
```suggestion
* Domain and central instances *MUST* ensure that the information within their databases is accurate and up to date and reconcile any differences periodically using the client and replication APIs described in this specification.
```

File: up-l3/udiscovery/v3/service.adoc [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+--
+
+[.specitem,oft-sid="dsn~discovery-data-reconciliation-frequency~1",oft-needs="impl,test"]
+--
+* *MUST* provide an implementation of data reconciliation using the above APIs where the frequency of reconciliation is customizable so the deployment can adjust the frequency based on the deployment needs.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:219]:
```suggestion
* Domain and central instances *MUST* provide an implementation of data reconciliation using the above APIs where the frequency of reconciliation is customizable so the deployment can adjust the frequency based on the deployment needs.
```

File: up-l3/udiscovery/v3/service.adoc [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+--
+
+[.specitem,oft-sid="dsn~discovery-data-reconciliation-data~1",oft-needs="impl,test"]
+--
+* When reconciling data, the data fetched *MUST* be used in lieu of current cached data.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:224]:
```suggestion
* When reconciling data, a client *MUST* replace corresponding data in its local cache with the fetched data. 
```

File: up-l3/udiscovery/v3/service.adoc [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+--
+
+[.specitem,oft-sid="dsn~discovery-data-reconciliation-data~1",oft-needs="impl,test"]
+--
+* When reconciling data, the data fetched *MUST* be used in lieu of current cached data. 
+--
+
+[.specitem,oft-sid="dsn~discovery-data-reconciliation-not-found~1",oft-needs="impl,test"]
+--
+* If `GetServiceTopics()` returns `UCode.NOT_FOUND` for data we had in our local instance, we *MUST* flush the local cache and then replicate the deletion. we fetch information to reconcile the data If the reconciliation fails, the service *MUST* log the error and continue to use the current cached data.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:229]: I am not sure who is `we` in these sentences. Can you rephrase this being (super-)explicit about the roles?

Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:229]: is this the `FindServiceTopics()` operation described above in the client API section?

File: up-l3/udiscovery/v3/service.adoc [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+[mermaid]
+ifdef::env-github[[source,mermaid]]
+----
+sequenceDiagram
+
+participant Local-UDiscovery
+participant Domain-UDiscovery
+participant Central-UDiscovery
+
+Domain-UDiscovery ->> Local-UDiscovery: GetServiceTopics()
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:245]: `FindServiceTopics()`?

**Review by sophokles73 — CHANGES_REQUESTED** [review]

File: up-l3/udiscovery/v3/service.adoc (hunk 17 above)
Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc:122]: FindServicesRequest does not include a ttl, there is only the uAttributes ttl for the request.

File: up-l3/udiscovery/v3/service.adoc (hunk 18 above)
Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc:127]: @sophokles73  yes, do I need to make it more explicit?

File: up-l3/udiscovery/v3/service.adoc [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+Domain-UDiscovery -->> Local-UDiscovery: FindServicesResponse
+Local-UDiscovery -->> Client: FindServicesResponse
+----
+
+
+== Replication API
+
+
+=== SetServiceTopics()
+The `SetServiceTopics()` API is used by the uDiscovery business logic to add, update, or remove information recursively between the local, domain, and central instances. The API is passed `SetServiceTopicsRequest` containing a repeated list of `UServiceTopicInfo` and `ttl` for when the info will expire. The function returns `SetServiceTopicsResponse` message if the method was successfully processed or it returns `UStatus` containing the error values specified below.
```
[the remark this answers is not in the document]
Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc:151]: I would assume it would be some central uE that is reponsible for handling software updates.

File: up-l3/udiscovery/v3/service.adoc (hunk 22 above)
Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc]: Cleaned up the requirements.

Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc]: Removed the ambiguity

File: up-l3/udiscovery/v3/service.adoc (hunk 23 above)
Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc:246]: The API is actually GetServiceTopics(), sorry for the confusion. I fixed this everywhere

**Review by stevenhartley — COMMENTED** [review]
feedback from @sophokles73 review

File: up-l3/udiscovery/v3/service.adoc [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+
+
+.Replication API Design
+[#udiscovery-replication-design]
+[mermaid]
+ifdef::env-github[[source,mermaid]]
+----
+sequenceDiagram
+
+OTAClient ->> Local-UDiscovery: SetServiceTopics()
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:197]: OTAClient?

File: up-l3/udiscovery/v3/service.adoc [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+----
+
+
+== Implementation Consideration
+
+It is possible the information in domain or central instances might become out of sync with what is stored in the other instances. In order to rectify this situation:
+
+[.specitem,oft-sid="dsn~discovery-data-reconciliation~1",oft-needs="impl,test"]
+--
+* Domain and central instances *MUST* ensure that the information within their databases is accurate and up to date and reconcile any differences periodically using the client and replication APIs described in this specification.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:214]: is the replication always only happening from _root_ to _leaf_, i.e. it is always initiated by a node asking its parent, never the other way around? This would also mean that nodes are only required to know their parent, but not the other way around, which would be good FMPOV.

Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc:205]: In an ideal world perhaps yes but due to the temporal nature of the leaves, it could be possible that the parents are not aware of changes in the leaves and they need to query the leaf (ex. the vehicle device does a factory reset of SW settings and there is no connectivity for a period of time. The replication needs to work in both directions to handle corner cases.

File: up-l3/udiscovery/v3/service.adoc (hunk 16 above)
Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc:106]: @sophokles73 do you want to rename the field `search_central` then to be something more like `recursive`?

**Review by stevenhartley — COMMENTED** [review]
Please see comments below

File: up-l3/udiscovery/v3/service.adoc [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+
+== Replication API
+
+
+=== SetServiceTopics()
+The `SetServiceTopics()` API is used by the uDiscovery business logic to add, update, or remove information recursively between the local, domain, and central instances. The API is passed `SetServiceTopicsRequest` containing a repeated list of `UServiceTopicInfo` and `ttl` for when the info will expire. The function returns `SetServiceTopicsResponse` message if the method was successfully processed or it returns `UStatus` containing the error values specified below. 
+
+[.specitem,oft-sid="dsn~discovery-replication-client~1",oft-needs="impl"]
+--
+* link:../languages.adoc[uProtocol Language Libraries] *MUST NOT* implement the client-side of {replicator-proto-ref} API as these APIs are only used internally by uDiscovery business logic.
```
[the remark this answers is not in the document]
Comment (sophokles73) [reply up-l3/udiscovery/v3/service.adoc:146]: Me neither, I therefore suggest removing the specitem definition.

File: up-l3/udiscovery/v3/service.adoc (hunk 26 above)
Comment (sophokles73) [reply up-l3/udiscovery/v3/service.adoc:205]: Couldn't we instead specify that child nodes **MUST** (re-)sync with their parent node whenever they start up?

File: up-l3/udiscovery/v3/service.adoc (hunk 27 above)
[the remark this answers is not in the document]
Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc:146]: ok done (when pushed)

File: up-l3/udiscovery/v3/service.adoc (hunk 26 above)
Comment (stevenhartley) [reply up-l3/udiscovery/v3/service.adoc:205]: @sophokles73 Ok I changed this, please take a look.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 28]
```diff
@@ -25,332 +24,93 @@ option java_multiple_files = true;
...
+// or all topics for all services for a given device (using wildcard ue_id and resource_id)
+message GetServiceTopicsRequest {
+  // The URI of the topics that we would like to query.
+  uprotocol.v1.UUri topic = 1;
 
-// Delete one or more nodes request
-message DeleteNodesRequest {
-  repeated string uris = 1;   // uProtocol formatted URI
+  // Recursively search up to the central level when true, otherwise only search up to the domain level
+  bool search_central = 2;
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:76]:
```suggestion
  // Recursively search up to the root node of the uDiscovery tree
  bool recursive = 2;
```

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 29]
```diff
@@ -25,332 +24,93 @@ option java_multiple_files = true;
...
   uprotocol.v1.UUri uri = 1;
-}
 
-
-// Response message returned from ResolveUri() API containing the resolved UUri
-message ResolveUriResponse {
-  // Resolved URI
-  uprotocol.v1.UUri uri = 1;
+  // Recursively search up to the central level when true, otherwise only search up to the domain level
+  bool search_central = 2;
```
Comment (sophokles73) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:108]:
```suggestion
  // Recursively search up to the root node of the uDiscovery tree
  bool recursive = 2;
```

File: up-l3/udiscovery/v3/client.adoc [hunk 30]
```diff
@@ -0,0 +1,128 @@
...
+The client facing APIs of uDiscovery provide the means for uEs to discover services (where they are located, versions, etc...), and topic metadata that a service produces (method id and permissions, publish topic message formats, etc...). The API is declared in the {discovery-proto-ref}.
+
+[.specitem,oft-sid="dsn~discovery-client-apis~1",oft-needs="impl"]
+--
+Each link:../languages.adoc[uProtocol Language Library] *MUST* implement the client-side APIs of {discovery-proto-ref} using the {up-l2-ref} RpcClient. 
+--
+
+=== FindServices()
+
+The FindServices() API is used to find the authority(ies), instance(s), and version(s) of services based on the passed URI. The only mandatory portion of the URI that must be passed is the ue_id, the rest can be discovered. When you pass the wildcard for authority_name, the search will be done up to the domain, in order to search in the central discovery database, you must set `search_central=true`.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/client.adoc:54]:
```suggestion
The FindServices() API is used to find service instances matching search criteria like authority, instance ID and (major) version. The only mandatory part of the provided (search) URI is the uEntity ID (`ue_id`), the rest can be discovered. The `recursive` flag can be used to include the uDiscovery service instance's parent nodes in the search.
```

File: up-l3/udiscovery/v3/client.adoc [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+
+[.specitem,oft-sid="dsn~discovery-findservices-error-notfound~1",oft-needs="impl,test"]
+--
+* *MUST* return a `UCode.NOT_FOUND` if no service matching the passed URI was found. 
+--
+
+[.specitem,oft-sid="dsn~discovery-findservices-error-invalid-argument~1",oft-needs="impl,test"]
+--
+* *MUST* return a `UCode.INVALID_ARGUMENT` if: 
+  * `search_central==true` and `authority_name` does not contain the wildcard `*`.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/client.adoc:64]:
```suggestion
  * `recursive==true` and `authority_name` does not contain the wildcard `*`.
```

File: up-l3/udiscovery/v3/client.adoc [hunk 32]
```diff
@@ -0,0 +1,128 @@
...
+--
+
+==== Example Use Cases
+
+ Below are example string URIs that could be passed to the API.
+
+.FindServices() Examples
+[#findservices-examples, cols="1,1,3", options="header", width="80%"]
+|===
+| URI | `search_central` | Description
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/client.adoc:81]:
```suggestion
| URI | `recursive` | Description
```

File: up-l3/udiscovery/v3/client.adoc [hunk 33]
```diff
@@ -0,0 +1,128 @@
...
+
+| `up:/FFFF1234/FF/0` | false | Find local instances & versions of service 1234.
+| `up://*/FFFF1234/FF/0` | false | Find instances & versions of service 1234 within a domain.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/client.adoc:84]:
```suggestion
| `up://*/FFFF1234/FF/0` | false | Find instances & versions of service 1234 that are known to the local uDiscovery service instance.
```

File: up-l3/udiscovery/v3/client.adoc [hunk 34]
```diff
@@ -0,0 +1,128 @@
...
+| `up://*/FFFF0004/FF/0` | true | Find instances & versions of service 4 that are globally addressable.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/client.adoc:85]:
```suggestion
| `up://*/FFFF0004/FF/0` | true | Find instances & versions of service 4 that are known to the local uDiscovery service instance or any of its parent nodes.
```

File: up-l3/udiscovery/v3/service.adoc [hunk 35]
```diff
@@ -0,0 +1,243 @@
...
+
+=== Recursive Lookups
+
+Recursive lookups happen when the client sets the following information in the request message of the FindService() or GetServiceTopics() API:
+[.specitem,oft-sid="dsn~discovery-recursive-authority-wildcard~1",oft-needs="impl,test"]
+--
+*  `authority_name` *MUST* contain the wildcard `*`.
+--
+
+If `search_central` inside of the `FindServicesRequest` or `GetServiceTopicsRequest` is `true``:
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:103]:
```suggestion
If the `recursive` flag of a `FindServicesRequest` or `GetServiceTopicsRequest` is set to `true`:
```

File: up-l3/udiscovery/v3/service.adoc [hunk 36]
```diff
@@ -0,0 +1,243 @@
...
+[.specitem,oft-sid="dsn~discovery-recursive-central~1",oft-needs="impl,test"]
+--
+* *MUST* search recursively up to and including the central instance.
```
Comment (sophokles73) [inline up-l3/udiscovery/v3/service.adoc:106]:
```suggestion
* *MUST* search recursively along the branch within the uDiscovery service instance hierarchy up to and including the root node of the tree.
```

**Review by sophokles73 — COMMENTED** [review]
It feels a little strange to have a separate `client.adoc` which basically replicates the information from `service.adoc`. This will also make it harder to keep everything consistent ...

**Review by sophokles73 — APPROVED** [review]
LGTM

Let's start with this and improve iteratively ...
